ledger/simulation/resources.go

```diff
@@ -0,0 +1,395 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package simulation
+
+import (
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// BoxRef is a reference to an application's box.
+type BoxRef struct {
+	// The actual ID of the app that owns the box.
+	App basics.AppIndex
+	// The name of the box.
+	Name string
+}
+
+// ResourceAssignment calculates the additional resources that a transaction or group could use, and
+// it tracks any referenced unnamed resources that fit within those limits.
+type ResourceAssignment struct {
+	Accounts    map[basics.Address]struct{}
+	MaxAccounts int
+
+	Assets    map[basics.AssetIndex]struct{}
+	MaxAssets int
+
+	Apps    map[basics.AppIndex]struct{}
+	MaxApps int
+
+	Boxes    map[BoxRef]struct{}
+	MaxBoxes int
+
+	MaxTotalRefs int
+}
+
+func makeTxnResourceAssignment(txn *transactions.Transaction, proto *config.ConsensusParams) ResourceAssignment {
+	if txn.Type != protocol.ApplicationCallTx {
+		return ResourceAssignment{}
+	}
+	return ResourceAssignment{
+		MaxAccounts:  proto.MaxAppTxnAccounts - len(txn.Accounts),
+		MaxAssets:    proto.MaxAppTxnForeignAssets - len(txn.ForeignAssets),
+		MaxApps:      proto.MaxAppTxnForeignApps - len(txn.ForeignApps),
+		MaxBoxes:     proto.MaxAppBoxReferences - len(txn.Boxes),
+		MaxTotalRefs: proto.MaxAppTotalTxnReferences - len(txn.Accounts) - len(txn.ForeignAssets) - len(txn.ForeignApps) - len(txn.Boxes),
+	}
+}
+
+func makeGlobalResourceAssignment(perTxnResources []ResourceAssignment) ResourceAssignment {
+	var globalResources ResourceAssignment
+	for i := range perTxnResources {
+		globalResources.MaxAccounts += perTxnResources[i].MaxAccounts
+		globalResources.MaxAssets += perTxnResources[i].MaxAssets
+		globalResources.MaxApps += perTxnResources[i].MaxApps
+		globalResources.MaxBoxes += perTxnResources[i].MaxBoxes
+		globalResources.MaxTotalRefs += perTxnResources[i].MaxTotalRefs
+	}
+	return globalResources
+}
+
+// IsEmpty returns true if the assignment has no resources.
+// NOTE: this is different from the traditional Go idiom of checking if the object is the zero value.
+func (a *ResourceAssignment) IsEmpty() bool {
```

Comment:
can't you call it something else then?

Reply:
isVacant?  hasNoItems?  isBereft?

Reply:
I might flip the meaning and make it `HasResources`

Reply:
done in 031c7c8d2